src/deblock.rs

```diff
@@ -0,0 +1,439 @@
+            p3: i32, p2: i32, p1: i32, p0: i32, q0: i32, q1: i32, q2: i32, q3: i32) {
+    data[pitch]   = (p3*3 + p2*2 + p1 + p0 + q0 + (1<<2) >> 3) as u16;
+    data[pitch*2] = (p3*2 + p2 + p1*2 + p0 + q0 + q1 + (1<<2) >> 3) as u16;
+    data[pitch*3] = (p3 + p2 + p1 + p0*2 + q0 + q1 + q2 + (1<<2) >> 3) as u16;
+    data[pitch*4] = (p2 + p1 + p0 + q0*2 + q1 + q2 + q3 + (1<<2) >> 3) as u16;
+    data[pitch*5] = (p1 + p0 + q0 + q1*2 + q2 + q3*2 + (1<<2) >> 3) as u16;
+    data[pitch*6] = (p0 + q0 + q1 + q2*2 + q3*3 + (1<<2) >> 3) as u16;
+}
+
+fn deblock14(data: &mut [u16], pitch: usize,
+             p6: i32, p5: i32, p4: i32, p3: i32, p2: i32, p1: i32, p0: i32,
+             q0: i32, q1: i32, q2: i32, q3: i32, q4: i32, q5: i32, q6: i32) {
+    data[pitch]   = (p6*7 + p5*2 + p4*2 + p3 + p2 + p1 + p0 + q0 + (1<<3) >> 4) as u16;
+    data[pitch*2] = (p6*5 + p5*2 + p4*2 + p3*2 + p2 + p1 + p0 + q0 + q1 + (1<<3) >> 4) as u16;
+    data[pitch*3] = (p6*4 + p5 + p4*2 + p3*2 + p2*2 + p1 + p0 + q0 + q1 + q2 + (1<<3) >> 4) as u16;
+    data[pitch*4] = (p6*3 + p5 + p4 + p3*2 + p2*2 + p1*2 + p0 + q0 + q1 + q2 + q3 + (1<<3) >> 4) as u16;
+    data[pitch*5] = (p6*2 + p5 + p4 + p3 + p2*2 + p1*2 + p0*2 + q0 + q1 + q2 + q3 + q4 + (1<<3) >> 4) as u16;
+    data[pitch*6] = (p6 + p5 + p4 + p3 + p2 + p1*2 + p0*2 + q0*2 + q1 + q2 + q3 + q4 + q5 + (1<<3) >> 4) as u16;
+    data[pitch*7] = (p5 + p4 + p3 + p2 + p1 + p0*2 + q0*2 + q1*2 + q2 + q3 + q4 + q5 + q6 + (1<<3) >> 4) as u16;
+    data[pitch*8] = (p4 + p3 + p2 + p1 + p0 + q0*2 + q1*2 + q2*2 + q3 + q4 + q5 + q6*2 + (1<<3) >> 4) as u16;
+    data[pitch*9] = (p3 + p2 + p1 + p0 + q0 + q1*2 + q2*2 + q3*2 + q4 + q5 + q6*3 + (1<<3) >> 4) as u16;
+    data[pitch*10] = (p2 + p1 + p0 + q0 + q1 + q2*2 + q3*2 + q4*2 + q5 + q6*4 + (1<<3) >> 4) as u16;
+    data[pitch*11] = (p1 + p0 + q0 + q1 + q2 + q3*2 + q4*2 + q5*2 + q6*5 + (1<<3) >> 4) as u16;
+    data[pitch*12] = (p0 + q0 + q1 + q2 + q3 + q4*2 + q5*2 + q6*7 + (1<<3) >> 4) as u16;
+}   
+
+fn deblock_edge4<'a>(slice: &'a mut PlaneMutSlice<'a>, pitch: usize, stride: usize,
+                 blimit: u16, limit: u16, thresh: u16, bd: usize) {
+    let mut s = 0;
+    let data = slice.as_mut_slice();
+    for _i in 0..4 {
+        let p = &mut data[s..];
+        let p1 = p[0] as i32;
+        let p0 = p[pitch] as i32;
+        let q0 = p[pitch*2] as i32;
+        let q1 = p[pitch*3] as i32;
+        if (p1 - p0).abs() as u16 <= limit &&
+            (q1 - q0).abs() as u16 <= limit &&
+            (p0 - q0).abs() as u16 * 2 + (p1 - q1).abs() as u16 / 2 <= blimit {
+                deblock4(thresh, bd, p, pitch, p1, p0, q0, q1);
+            }
+        s += stride;
+    }
+}
+
+fn deblock_edge6<'a>(slice: &'a mut PlaneMutSlice<'a>, pitch: usize, stride: usize,
+                 blimit: u16, limit: u16, thresh: u16, bd: usize) {
+    let mut s = 0;
+    let flat = 1 << bd - 8;
+    let data = slice.as_mut_slice();
+    for _i in 0..4 {
```

Comment:
``` rust
for p in data.chunks_mut(stride).take(4) {
  assert!(p.len() >= pitch * 5);
```

Should have mir and llvm more sure about what to do. (also see [slice patterns](https://rust-lang-nursery.github.io/edition-guide/rust-2018/slice-patterns.html#slice-patterns)).